src/rgw/rgw_rest_pubsub.cc

```diff
@@ -885,24 +901,32 @@ void RGWPSSetTopicAttributesOp::execute(optional_yield y) {
     // initialize the persistent queue's location, using ':' as the namespace
     // delimiter because its inclusion in a TopicName would break ARNs
     dest.persistent_queue = string_cat_reserve(topic_arn.account, ":", topic_name);
-
-    op_ret = driver->add_persistent_topic(this, y, dest.persistent_queue);
-    if (op_ret < 0) {
-      ldpp_dout(this, 4)
-          << "SetTopicAttributes Action failed to create queue for "
-             "persistent topics. error:"
-          << op_ret << dendl;
-      return;
+    dest.num_shards = s->get_cct()->_conf.get_val<uint64_t>("rgw_bucket_persistent_notif_num_shards");
+    for(const auto& q: dest.get_shard_names()){
+      op_ret = driver->add_persistent_topic(this, y, q);
+      if (op_ret < 0) {
+        ldpp_dout(this, 4)
+            << "SetTopicAttributes Action failed to create queue for "
+              "persistent topics. error:"
+            << op_ret << dendl;
+        return;
+      }
     }
+    ldpp_dout(this, 20) << "Successfully created " <<dest.num_shards<<" shards for topic: " + topic_name<<dendl;
+
   } else if (already_persistent && !topic_needs_queue(dest)) {
     // changing the persistent topic to non-persistent.
-    op_ret = driver->remove_persistent_topic(this, y, result.dest.persistent_queue);
-    if (op_ret != -ENOENT && op_ret < 0) {
-      ldpp_dout(this, 4) << "SetTopicAttributes Action failed to remove queue "
-                            "for persistent topics. error:"
-                         << op_ret << dendl;
-      return;
+    dest.num_shards = 0; // set to 0 to indicate no sharding
```

Comment:
i think you should set that after the call to `get_shard_names()` or you would get an empty list?